CRM/Admin/Form/Setting/Localization.php

```diff
@@ -354,11 +354,13 @@ public static function onChangeDefaultCurrency($oldCurrency, $newCurrency, $meta
    * @return array
    */
   public static function getDefaultLanguageOptions() {
-    return [
+    $availableOptions = [
       '*default*' => ts('Use default site language'),
       'undefined' => ts('Leave undefined'),
       'current_site_language' => ts('Use language in use at the time'),
     ];
+    $availableLanguages = array_merge($availableOptions, CRM_Core_I18n::languages());
```

Comment:
wouldn't it make sense to restrict those to the enabled ones - i.e. ``CRM_Core_I18n::languages(true)``? Hadn't tested this, though.

Reply:
Point is, idk how the system behaves when a contact's current preferred language is not enabled - that could cause issues e.g. when editing.

Reply:
Well, to be frank, I think that this line:
`$availableLanguages = array_merge($availableOptions, CRM_Core_I18n::languages());`
should become
`$availableLanguages = array_merge($availableOptions, getDefaultLocaleOptions());`
Am I correct ? If so, I will make a new commit to do it as i wrote above.

Now, in regards to the behaviour, while I don't understand how a contact's preferred language cannot be enabled, I will assume that you talk about having the preferred language set to NULL and then someone edits that contact,so I did a couple of test:

### Vanilla CiviCRM (without patch):
Editing a contact without a language preference (set to NULL), it prepopulates his/her 'Preferred communication language'  to the `Default Language` (in my case English)
Localization settings:

![image](https://user-images.githubusercontent.com/14973113/117570528-5d46d800-b0c2-11eb-9163-021026f0ffba.png)

### Patched CiviCRM:
A. Editing a contact without a language preference (set to NULL), it prepopulates his/her 'Preferred communication language'  to the language defined in the 'Default Language for contacts' , which I set to 'Czech' and actually does make sense.

Localization settings:

![image](https://user-images.githubusercontent.com/14973113/117570741-8ae05100-b0c3-11eb-9fee-2c8633aa4ecd.png)

B. Editing a contact that has already a preferred comm language (eg. English), retains the language preconfigured for that contact without altering it.

Were you referring to something else that I missed to see ?

Reply:
@VangelisP I don't think we should be altering a preferred language from NULL (ie unknown) on edit- on new we should use the default but if it is already NULL we should respect that

Reply:
@eileenmcnaughton just to be clear, I'm stating the **current** default behaviour (without or with the patch), as per [CRM-7119](https://issues.civicrm.org/jira/browse/CRM-7119) -> [source](https://github.com/civicrm/civicrm-core/blob/master/CRM/Contact/Form/Edit/CommunicationPreferences.php#L146-L148). In fact, I've used that assumption as a basis and I've worked on top of that. 

If you think that this not the intended behaviour, then we need to remove CRM-7119's addition and introduce on the form's `setDefaults` a way to identify if the contact being edited is a new or existing contact and then work some conditionals there. Correct me if I'm wrong please.

Reply:
@VangelisP that second part seems right - but it looks like ```CRM_Core_I18n::getContactDefaultLanguage()``` already falls back to lcMessages

```
  /**
   * Get the default language for contacts where no language is provided.
   *
   * Note that NULL is a valid option so be careful with checking for empty etc.
   *
   * NULL would mean 'we don't know & we don't want to hazard a guess'.
   *
   * @return string
   */
  public static  function getContactDefaultLanguage() {
    $language = Civi::settings()->get('contact_default_language');
    if ($language == 'undefined') {
      return NULL;
    }
    if (empty($language) || $language === '*default*') {
      $language = civicrm_api3('setting', 'getvalue', [
        'name' => 'lcMessages',
        'group' => CRM_Core_BAO_Setting::LOCALIZATION_PREFERENCES_NAME,
      ]);
    }
    elseif ($language == 'current_site_language') {
      return CRM_Core_I18n::getLocale();
    }

    return $language;
  }
```